modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersValidator.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.http.netty4;
+
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.DefaultHttpRequest;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMessage;
+import io.netty.handler.codec.http.HttpRequest;
+
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.TriConsumer;
+import org.elasticsearch.http.HttpHeadersValidationException;
+import org.elasticsearch.http.HttpPreRequest;
+import org.elasticsearch.http.netty4.HttpHeadersValidator.ValidatableHttpHeaders.ValidationResult;
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.getHttpHeadersAsMap;
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.translateRequestMethod;
+
+/**
+ * Provides utilities used to hook into the netty pipeline and run a validation check on each HTTP request's headers.
+ * See also {@link Netty4HttpHeaderValidator}.
+ */
+public final class HttpHeadersValidator {
+
+    /**
+     * An async HTTP headers validator function that receives as arguments part of the incoming HTTP request
+     * (except the body contents, see {@link HttpPreRequest}), as well as the netty channel that the request is
+     * being received over, and must then call the {@code ActionListener#onResponse} method on the listener parameter
+     * in case the validation is to be considered successful, or otherwise call {@code ActionListener#onFailure}.
+     */
+    private final TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResult>> validator;
```

Comment:
Can we use a proper class with a named method instead of a TriConsumer here (and elsewhere) ? Also, I don't think we need to skirt around the fact that this doing authentication (it makes the code harder to understand about). I think we should keep `Netty4HttpHeaderValidator` named as is, but have it accept a HttpAuthenticator as a constructor param.

Reply:
> I think we should keep Netty4HttpHeaderValidator named as is, but have it accept a HttpAuthenticator as a constructor param.

I not convinced this will achieve more readable code. I slightly prefer we be consistent and only use "validator" in the `transport-netty4` plugin. For example, after addressing your other suggestions, the code here looks like:

```
@FunctionalInterface
public interface Validator extends TriConsumer<HttpPreRequest, Channel, ActionListener<ThreadContext.StoredContext>> {}

public static Netty4HttpHeaderValidator getValidatorInboundHandler(Validator validator) {
....
}
```
I feel like "authenticator" would add another concept to wrestle with.

Reply:
We've discussed this separately, and I was convinced to rename most "validation" things to "authentication" things, except the `Netty4HttpHeaderValidator`. I've pushed [625e56c](https://github.com/elastic/elasticsearch/pull/95112/commits/625e56c38b302bd5f7e788ad18f0e183e5c40f1a) .